### PR TITLE
[issue 195] terminate streams gracefully

### PR DESCRIPTION
**Pull request by moh-osman3**
Resolves https://github.com/f5/otel-arrow-adapter/issues/195

This PR solves an issue with stream termination, where once `max_connection_age` is reached in the server (otlpreceiver), it will shutdown and return a `NO_ERROR` error to the client. To resolve this, the client (otlpexporter) will set a `max_stream_lifetime` in the config to signal a graceful shutdown in the client before the server reaches the `max_connection_age`

For testing I ran the arrow collector with a `record.yaml` which had a telemetry generator receiver -> otlpexporter -> otlpreceiver -> fileexporter pipeline, and confirmed I was no longer seeing the `arrow stream shutdown` error

### CODE REVIEW

File: collector/gen/exporter/otlpexporter/internal/arrow/stream.go [hunk 1]
```diff
@@ -325,15 +346,23 @@ func (s *Stream) read(_ context.Context) error {
 	// cancel a call to Recv() but the call to processBatchStatus
 	// is non-blocking.
 	for {
+		// Note: a dead stream error might be possible if the client
+		// calls CloseSend() and is stuck waiting for a response from the server.
 		resp, err := s.client.Recv()
 		if err != nil {
+			// Once the send direction of stream is closed the server should
+			// return an error that mentions an EOF.
+			if status, ok := status.FromError(err); ok && status.Message() == "EOF" {
```
Comment (jmacd) [inline collector/gen/exporter/otlpexporter/internal/arrow/stream.go:355]: Is there a `status.Code()` we can test for, too?  
I'm hoping it's `codes.OK` but gRPC is hard to predict.

File: collector/gen/exporter/otlpexporter/internal/arrow/stream.go [hunk 2]
```diff
@@ -325,15 +346,23 @@ func (s *Stream) read(_ context.Context) error {
 	// cancel a call to Recv() but the call to processBatchStatus
 	// is non-blocking.
 	for {
+		// Note: a dead stream error might be possible if the client
+		// calls CloseSend() and is stuck waiting for a response from the server.
```
Comment (jmacd) [inline collector/gen/exporter/otlpexporter/internal/arrow/stream.go:350]:
```suggestion
		// Note: if the client has called CloseSend() and is waiting for a response from the server.
		// If the server fails for some reason, we will wait until some other condtion, such as a context
		// timeout.  TODO: possibly, improve to wait for no outstanding requests and then stop reading.
```

File: collector/gen/exporter/otlpexporter/internal/arrow/stream.go (hunk 1 above)
Comment (moh-osman3) [reply collector/gen/exporter/otlpexporter/internal/arrow/stream.go]: Thanks for the review! Hmm sadly it seems that in this specific case, `status.Code() == codes.Unknown`, so I've now added a check to check the message and code

File: collector/gen/exporter/otlpexporter/internal/arrow/stream.go (hunk 2 above)
Comment (moh-osman3) [reply collector/gen/exporter/otlpexporter/internal/arrow/stream.go]: Updated the comment!

**Review by jmacd — APPROVED** [review]

**Pull request merged**